### PR TITLE
windows: build shared libs too

### PR DESCRIPTION
**Pull request by dougalsutherland**
Currently the Windows build only makes static libraries, which doesn't work so well for consumers that want to build shared libs (e.g. https://github.com/conda-forge/shogun-cpp-feedstock/pull/19). (Unix builds currently make both static and shared.)

This PR makes it build both static and shared libs, modeled after how snappy and a few other feedstocks do it.

(Also adds me as a maintainer, because why not.)

### CODE REVIEW

**Comment by jakirkham** [comment]
Would it be possible to send the patch upstream? Or have you already?

**Comment by MaxRis** [comment]
Looks good, thanks!

**Comment by dougalsutherland** [comment]
Sure, I'll make a version that makes it optional and send it upstream.

**Comment by dougalsutherland** [comment]
Apparently, when you build protobuf as a shared library on Windows, you're supposed to then compile with `PROTOBUF_USE_DLLS`. I _think_ that maybe the cmake files don't do this properly for the test targets. Trying adding a patch to make them do this; if this works, I'll work on a fuller solution and send it upstream.

(This would be a lot easier to iterate on if I had a Windows machine to run this stuff myself on...maybe I'll look into setting up a VM.)

**Comment by jakirkham** [comment]
I think @msarahan had some Vagrant files that would be useful for this. Of course I’m recalling this from two years ago. So no idea what state they are in or if you should be using something else instead.

**Comment by MaxRis** [comment]
@dougalsutherland I can try on my side with Windows related things. But not get what the issue currently happens with test targets on Windows.

**Comment by dougalsutherland** [comment]
That apparently wasn't it. I'm confused, and not looking forward to iterating on the AppVeyor queue. I'll look into whether I can get a copy of Windows I can run in a VM through work...sigh.

**Comment by jakirkham** [comment]
Might take a look at PR ( https://github.com/ContinuumIO/vagrant-images/pull/1 ). There's a Vagrantfile for building a Windows VM image. IIRC it uses a freely available copy of Windows, but has a timeout after which one must rebuild the VM from scratch. Would ask @msarahan for more details

File: recipe/bld.bat [hunk 1]
```diff
@@ -4,29 +4,57 @@ if errorlevel 1 exit 1
...
-cd build
+cd build-shared
 if errorlevel 1 exit 1
-mkdir release
+
+cmake -G "NMake Makefiles" ^
+         -DCMAKE_BUILD_TYPE=Release ^
+         -DCMAKE_PREFIX_PATH="%LIBRARY_PREFIX%" ^
+         -DCMAKE_INSTALL_PREFIX="%LIBRARY_PREFIX%" ^
+         -DBUILD_SHARED_LIBS=ON ^
```
Comment (jakirkham) [inline recipe/bld.bat]: Think we need `-Dprotobuf_BUILD_SHARED_LIBS=ON`.

File: recipe/bld.bat [hunk 2]
```diff
@@ -4,29 +4,57 @@ if errorlevel 1 exit 1
...
+         -Dprotobuf_WITH_ZLIB=ON ^
```
Comment (jakirkham) [inline recipe/bld.bat]: Also think we need `-Dprotobuf_MSVC_STATIC_RUNTIME=OFF`.

**Comment by jakirkham** [comment]
Probably already have seen this, but if not [this README]( https://github.com/google/protobuf/blob/v3.5.2/cmake/README.md ) looks pretty useful. The suggestions above were pulled from that README.

File: recipe/bld.bat (hunk 1 above)
Comment (dougalsutherland) [reply recipe/bld.bat]: `BUILD_SHARED_LIBS` [sets that too](https://github.com/google/protobuf/blob/v3.5.2/cmake/CMakeLists.txt#L17).

File: recipe/bld.bat (hunk 2 above)
Comment (dougalsutherland) [reply recipe/bld.bat]: This one [doesn't do anything](https://github.com/google/protobuf/blob/v3.5.2/cmake/CMakeLists.txt#L24) when `BUILD_SHARED_LIBS` is set. :/

**Comment by nehaljwani** [comment]
Ping! Would be nice to have this. The packages on the main channel for windows have only shared libs and the package in conda forge seems to have only static lib. This is causing trouble :(

Also, see:
https://github.com/google/protobuf/issues/4118

**Comment by jakirkham** [comment]
Would you like to work on this PR, @nehaljwani? Would be happy to have your help here. :)

**Comment by nehaljwani** [comment]
I have sent a PR to google at https://github.com/google/protobuf/pull/4854

If that gets merged, I'll try building a project which links to the DLL (most likely the newer version of mysql-connector-python) and if everything goes good, then I'll wait for them to make a release, or backport the changes and make a PR to this feedstock.

**Comment by wesm** [comment]
Closed by #22

**Pull request closed without merging**